### PR TITLE
fix(screentracker): prevent Send() from blocking when ReadyForInitialPrompt stays false

### LINKED ISSUE #209: Send() blocks indefinitely when ReadyForInitialPrompt returns false despite Status() reporting "stable"
## Bug

`POST /message` with `type: "user"` hangs indefinitely when `ReadyForInitialPrompt` never returns `true` for the current screen, even though `GET /status` returns `"stable"`.

## Root cause

`statusLocked()` does not check `initialPromptReady`. It returns `"stable"` when the screen is stable and the queue is empty. `Send()` passes the status validation and enqueues the message. But the `stableSignal` (which the send loop waits on) requires `initialPromptReady` to be `true`. If `ReadyForInitialPrompt` never returns `true`, the signal never fires and `Send()` blocks forever.

```
statusLocked() = "stable"   →  Send() passes status check, enqueues
stableSignal requires:       →  initialPromptReady && queue > 0 && screenStable
initialPromptReady = false   →  signal never fires
Send() blocks on errCh       →  forever
```

## Real-world trigger

Claude Code v2.1.87 shows a theme selection onboarding screen that uses `╌╌╌` (Unicode dashed lines) instead of `───` (box-drawing characters). The message box detection in `findGreaterThanMessageBox` / `findGenericSlimMessageBox` fails because it looks for `───────────────` and `>` which aren't present on this screen. `initialPromptReady` stays `false` and all `user`-type messages hang.

Last 6 lines of the actual screen content:
```
  1  function greet() {                                                         
  2 -  console.log("Hello, World!");                                            
  2 +  console.log("Hello, Claude!");                                           
  3  }                                                                          
 ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ 
  Syntax theme: Monokai Extended (ctrl+t to disable)                            
```

Note: `╌` (U+254C, BOX DRAWINGS LIGHT DOUBLE DASH HORIZONTAL) vs `─` (U+2500, BOX DRAWINGS LIGHT HORIZONTAL).

## Reproducing test

```go
func TestSendBlocksWhenInitialPromptNotReady(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	t.Cleanup(cancel)

	mClock := quartz.NewMock(t)
	agent := &testAgent{screen: "onboarding screen without message box"}
	cfg := st.PTYConversationConfig{
		Clock:                 mClock,
		SnapshotInterval:      100 * time.Millisecond,
		ScreenStabilityLength: 200 * time.Millisecond,
		AgentIO:               agent,
		ReadyForInitialPrompt: func(screen string) bool {
			return false // Simulates failed message box detection.
		},
		Logger: slog.New(slog.NewTextHandler(io.Discard, nil)),
	}
	c := st.NewPTY(ctx, cfg, &testEmitter{})
	c.Start(ctx)

	// Fill snapshot buffer to reach stability.
	advanceFor(ctx, t, mClock, 300*time.Millisecond)

	// Status reports "stable" — Send() will pass the status check.
	assert.Equal(t, st.ConversationStatusStable, c.Status())

	// But Send() blocks forever because stableSignal requires
	// initialPromptReady, which is false.
	var sendDone atomic.Bool
	go func() {
		_ = c.Send(st.MessagePartText{Content: "hello"})
		sendDone.Store(true)
	}()
	advanceFor(ctx, t, mClock, 20*time.Second)

	assert.False(t, sendDone.Load(),
		"Send() blocks forever: status said stable but stableSignal never fires")
}
```

## Possible fixes

1. **Make `statusLocked()` check `initialPromptReady`**: return `"initializing"` or `"changing"` when `initialPromptReady` is false. This would cause `Send()` to reject the message with `ErrMessageValidationChanging` instead of hanging.
2. **Improve message box detection**: add `╌╌╌` as an alternative to `───` in `findGreaterThanMessageBox` / `findGenericSlimMessageBox`.
3. **Both**: fix the detection AND add the safety check in `statusLocked()`.

Option 3 is likely the right approach — fix the detection for this specific Claude Code version, and add the statusLocked guard so future detection failures fail fast instead of hanging.

Discovered while smoke-testing #208.

> 🤖 This response was generated by Coder Agents.

### PR DESCRIPTION
**Pull request by johnstcn**
Fixes #209. Stacked on #208.

## Changes

- Make `statusLocked()` return `"changing"` when `initialPromptReady` is false, so `Send()` rejects with `ErrMessageValidationChanging` instead of blocking forever
- Add `╌` (U+254C) as alternative to `─` (U+2500) in `findGreaterThanMessageBox` / `findGenericSlimMessageBox` for Claude Code v2.1.87 onboarding screens
- Add testdata fixture for the dashed-line Claude screen
- Add `TestSendRejectsWhenInitialPromptNotReady` regression test
- Update four existing tests that asserted `stable` when `initialPromptReady` was false

<details><summary>Implementation plan and decision log</summary>

### Root cause

`statusLocked()` didn't check `initialPromptReady`. It returned `"stable"` when the screen was calm, but the `stableSignal` (which the send loop waits on) required `initialPromptReady == true`. If readiness never arrived, `Send()` blocked forever.

### Key decisions

| Decision | Rationale |
|----------|----------|
| Return `"changing"` (not `"initializing"`) | Snapshot buffer is full (past initializing). `"changing"` matches `ErrMessageValidationChanging`. |
| Unconditional guard (not only when InitialPrompt configured) | `stableSignal` gates ALL outbound messages on `initialPromptReady`, not just initial prompts. Status must reflect actual send capability. |
| U+254C only (not broader char set) | This is the specific character seen in the wild. The statusLocked guard is the safety net for future unknown characters. |
| Update existing test assertions | Four tests documented the buggy behavior (stable when not ready). Updated to expect `changing`. |

</details>

> 🤖 Written by a Coder Agent. Will be reviewed by a human.

### CODE REVIEW

File: lib/screentracker/pty_conversation.go [hunk 1]
```diff
@@ -571,6 +571,14 @@ func (c *PTYConversation) statusLocked() ConversationStatus {
 		return ConversationStatusChanging
 	}
 
+	// The send loop gates outbound messages on initialPromptReady.
+	// Report "changing" until readiness is detected so that Send()
+	// rejects with ErrMessageValidationChanging instead of blocking
+	// indefinitely on a stableSignal that will never fire.
+	if !c.initialPromptReady {
+		return ConversationStatusChanging
```
Comment (johnstcn) [inline lib/screentracker/pty_conversation.go:579]: **Obs** `ConversationStatusChanging` now carries two distinct meanings: "screen content is actively changing" and "agent readiness detection hasn't fired." *(Edge Case Analyst, Contract Auditor)*

> An external consumer watching `/status` cannot distinguish between transient screen instability (which resolves on its own) and permanent readiness failure (which may never resolve without user intervention).

This is strictly better than the old behavior (status said "stable" but `Send()` hung). Both `changing` and `initializing` map to `running` in the HTTP layer anyway, so the public API surface doesn't change. If a future feature needs to surface "agent not recognized" vs. "screen still settling," a new status value or a separate readiness field would be needed — but that's out of scope here.

File: lib/msgfmt/message_box.go [hunk 2]
```diff
@@ -5,14 +5,14 @@ import (
...
-// ───────────────
+// ───────────────  (or ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌)
 // >
-// ───────────────
+// ───────────────  (or ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌)
 // Used by Claude Code, Goose, and Aider.
 func findGreaterThanMessageBox(lines []string) int {
 	for i := len(lines) - 1; i >= max(len(lines)-6, 0); i-- {
 		if strings.Contains(lines[i], ">") {
-			if i > 0 && strings.Contains(lines[i-1], "───────────────") {
```
Comment (johnstcn) [inline lib/msgfmt/message_box.go:15]: **Obs** `findGenericSlimMessageBox` accepts mixed border styles — e.g. `───` on top and `╌╌╌` on bottom would match. *(Edge Case Analyst, Contract Auditor)*

> The top and bottom borders are checked independently with `||`. In practice, Claude Code uses the same character for both borders, so this is defensive rather than problematic.

Agreed this is fine — more permissive detection reduces false negatives, and the narrow scan window (last 9 lines) plus the prompt-character requirement on the middle line make false positives unlikely. Worth knowing if a false-positive concern arises later.

**Review by johnstcn — COMMENTED** [review]
🤖 Deep review of PR #215 (5 reviewers: Test Auditor, Edge Case Analyst, Contract Auditor, Concurrency Reviewer, Style Reviewer).

Core fix is clean — `statusLocked()` now reflects actual `Send()` capability, and the concurrency model is correct (one-way latch under lock, no new hazards). Good regression test that directly exercises the bug scenario.

Formatting issues (extra tabs, missing newline, stale doc comments) were caught by the review and already fixed in 934e832. Two observations remain worth noting inline. No blocking findings remain.

File: lib/msgfmt/message_box.go [hunk 3]
```diff
@@ -22,14 +22,14 @@ func findGreaterThanMessageBox(lines []string) int {
...
 // Usually something like
-// ───────────────
+// ───────────────  (or ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌)
 // |
-// ───────────────
+// ───────────────  (or ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌)
 func findGenericSlimMessageBox(lines []string) int {
 	for i := len(lines) - 3; i >= max(len(lines)-9, 0); i-- {
-		if strings.Contains(lines[i], "───────────────") &&
+		if (strings.Contains(lines[i], "───────────────") || strings.Contains(lines[i], "╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌")) &&
```
Comment (mafredri) [inline lib/msgfmt/message_box.go]: **P3** The `╌` addition in `findGenericSlimMessageBox` has no test coverage. The fixture (`msg_dashed_box.txt`) uses `>` as the prompt character, so `findGreaterThanMessageBox` matches first and this function is never reached. A fixture with `╌` borders and `│` or `❯` as the prompt indicator would close the gap. Without it, a typo in the dashed string on any of these three lines would pass the entire test suite, and the `statusLocked()` guard would mask the regression as permanent "running" status rather than a clear detection failure. *(Bisky P3, Mafuuu P3)*

Separately: `"╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌"` is now inlined 3 times (lines 15, 30, 32), mirroring the pre-existing 3x `"───────────────"`. A `containsHorizontalBorder(line string) bool` helper would collapse all six sites and prevent copy-paste miscounts on these visually-identical 15-char Unicode strings. *(Gon P3)*

> 🤖

File: docs/plans/issue-209-research.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+1. `statusLocked()` does NOT check `initialPromptReady`. It returns `"stable"` when the screen is stable and the queue is empty.
+2. `Send()` checks `statusLocked() != ConversationStatusStable` — this passes, so the message is enqueued.
+3. The send loop blocks on `<-c.stableSignal`.
+4. The snapshot loop only fires `stableSignal` when `c.initialPromptReady && len(c.outboundQueue) > 0 && c.isScreenStableLocked()`.
+5. Since `initialPromptReady` is `false` and never becomes `true`, the signal never fires.
+6. `Send()` blocks forever on `<-errCh`.
+
+**Real-world trigger:** Claude Code v2.1.87 shows a theme selection onboarding screen using `╌╌╌` (U+254C, BOX DRAWINGS LIGHT DOUBLE DASH HORIZONTAL) instead of `───` (U+2500, BOX DRAWINGS LIGHT HORIZONTAL). The message box detection fails because `findGreaterThanMessageBox` / `findGenericSlimMessageBox` look for `───────────────` which isn't present. `ReadyForInitialPrompt` stays `false`.
+
```
Comment (mafredri) [inline docs/plans/issue-209-research.md]: **P3** Line numbers for `pty_conversation.go` functions are wrong at the base SHA. `statusLocked()` is cited at 505 (actual: 549), `Send()` at 358 (actual: 379), snapshot loop signal at 229 (actual: 250). The offsets are non-uniform, ruling out a version mismatch. The code content in each snippet is accurate, so readers will find the right code by function name, but the specific line citations are fabricated precision. The smaller files (`message_box.go`, `agent_readiness.go`) have correct numbers, suggesting confabulation for the large file. *(Mafu-san, downgraded from P2: consequence is wasted reader time, not wrong conclusions.)*

> 🤖

File: lib/screentracker/pty_conversation_test.go [hunk 5]
```diff
@@ -1743,3 +1743,38 @@ func TestInitialPromptSent(t *testing.T) {
...
+	t.Cleanup(cancel)
+
+	mClock := quartz.NewMock(t)
+	agent := &testAgent{screen: "onboarding screen without message box"}
+	cfg := st.PTYConversationConfig{
+		Clock:                 mClock,
+		SnapshotInterval:      100 * time.Millisecond,
+		ScreenStabilityLength: 200 * time.Millisecond,
+		AgentIO:               agent,
+		ReadyForInitialPrompt: func(screen string) bool {
```
Comment (mafredri) [inline lib/screentracker/pty_conversation_test.go]: **Note** This is the only `ReadyForInitialPrompt` lambda using `screen` as the parameter name. All 17 others in this file (and the config field definition) use `message`. The caller at `pty_conversation.go:224` passes a variable called `screen`, so `screen` is arguably more correct, but creates a visual inconsistency. *(Gon)*

> 🤖

**Review by mafredri — COMMENTED** [review]
Correct fix, well-researched. The root cause analysis is genuine: `statusLocked()` returning "stable" while `stableSignal` required `initialPromptReady` was a real semantic mismatch, and the guard closes the gap at the right layer. Belt-and-suspenders approach (detection fix + status guard) is the right call for this class of bug.

Three P3s, two notes. The headline finding is a test coverage gap in `findGenericSlimMessageBox`'s dashed-border path, which two reviewers found independently. The `statusLocked()` guard and the `findGreaterThanMessageBox` path are well-tested.

As Hisoka put it: *"The API-visible change ('running' instead of 'stable' when `initialPromptReady` is false) is correct: it was lying before."*

Process note: the style-cleanup commit (`934e832`) fixes gofmt indentation and doc comments that should have been part of the earlier code commits, suggesting the initial commits were pushed without running the formatter.

> 🤖 This review was automatically generated with Coder Agents.

File: lib/screentracker/pty_conversation.go (hunk 1 above)
Comment (mafredri) [reply lib/screentracker/pty_conversation.go:579]: **Note** Adding to this thread: when readiness never fires, the system now reports "running" indefinitely with no log line or error explaining *why*. The user's only recourse is inspecting the terminal directly. This is strictly better than the old hang, and the dual-meaning limitation is acknowledged as out of scope here. Noting for awareness since there's no planned follow-up or tracking issue. *(Mafuuu)*

> 🤖

File: lib/msgfmt/message_box.go (hunk 3 above)
Comment (johnstcn) [reply lib/msgfmt/message_box.go]: 🤖 Both addressed in cb5b5ce:
- Extracted `containsHorizontalBorder()` helper — collapses all 6 inline checks
- Added `msg_slim_dashed_box.txt` fixture with `│` prompt to exercise `findGenericSlimMessageBox`'s `╌` path

File: lib/screentracker/pty_conversation_test.go (hunk 5 above)
Comment (johnstcn) [reply lib/screentracker/pty_conversation_test.go]: 🤖 Renamed `screen` → `message` for consistency in cb5b5ce.

File: docs/plans/issue-209-research.md (hunk 4 above)
Comment (johnstcn) [reply docs/plans/issue-209-research.md]: 🤖 Removed `docs/plans/` from the branch entirely in cb5b5ce.

File: lib/screentracker/pty_conversation_test.go [hunk 6]
```diff
@@ -1304,7 +1302,7 @@ func TestInitialPromptReadiness(t *testing.T) {
 		assert.Equal(t, st.ConversationStatusStable, c.Status())
 	})
 
-	t.Run("no initial prompt - normal status logic applies", func(t *testing.T) {
+	t.Run("no initial prompt - status is changing when readiness never fires", func(t *testing.T) {
```
Comment (mafredri) [inline lib/screentracker/pty_conversation_test.go]: **P3** This test name is hard to distinguish from the sibling at line 1332: "no initial prompt configured - normal status logic applies." Both omit `InitialPrompt`. The difference is that this test sets `ReadyForInitialPrompt` to always return false, while line 1332 omits it entirely (defaults to `return true`). "No initial prompt" reads as shorthand for "no initial prompt configured." Could we name it after the distinguishing condition, e.g. `"ReadyForInitialPrompt always false - status is changing"`? *(Gon)*

> 🤖

File: lib/screentracker/pty_conversation.go [hunk 7]
```diff
@@ -571,6 +571,14 @@ func (c *PTYConversation) statusLocked() ConversationStatus {
 		return ConversationStatusChanging
 	}
 
+	// The send loop gates outbound messages on initialPromptReady.
```
Comment (mafredri) [inline lib/screentracker/pty_conversation.go]: **Note** "The send loop gates outbound messages on initialPromptReady" misattributes the gating. `initialPromptReady` is checked by the snapshot loop (line 250), not the send loop. The send loop blocks on `stableSignal`, which the snapshot loop only fires when `initialPromptReady && len(outboundQueue) > 0 && isScreenStableLocked()`. The described effect is correct, but "The snapshot loop gates stableSignal on initialPromptReady" would be more precise. Same phrasing appears in the test comment at line 1326. *(Gon)*

> 🤖

**Review by mafredri — COMMENTED** [review]
All three round 1 P3s addressed cleanly in cb5b5ce: `containsHorizontalBorder()` helper extracted, `msg_slim_dashed_box.txt` fixture covers `findGenericSlimMessageBox`'s `╌` path, `docs/plans/` removed, parameter renamed. Both notes resolved.

One new P3 (test naming), one note (comment wording). The fix itself is solid. Four reviewers independently verified the `statusLocked()` guard, fixture coverage, and lifecycle correctness.

> 🤖 This review was automatically generated with Coder Agents.

File: lib/screentracker/pty_conversation_test.go (hunk 6 above)
Comment (johnstcn) [reply lib/screentracker/pty_conversation_test.go]: fixed

File: lib/screentracker/pty_conversation.go (hunk 7 above)
Comment (johnstcn) [reply lib/screentracker/pty_conversation.go]: fixed
